compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -64,6 +66,85 @@ static llvm::cl::opt<int> clInlineConstantByteLength(
 static const char kRootOpAttr[] = "__root_op__";
 static const char kFusionGroupsAttr[] = "__fused_op__";
 
+using namespace mlir;
+using namespace mlir::iree_compiler;
+using namespace mlir::iree_compiler::IREE;
+
+//===----------------------------------------------------------------------===//
+// Helpers for fusion group formation
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// A rewriter that keeps track of all tensor::DimOps.
+class TensorDimTrackingRewriter : public IRRewriter {
+ public:
+  /// Create a new rewriter: Scan the given op for tensor::DimOps.
+  TensorDimTrackingRewriter(Operation *op) : IRRewriter(op->getContext()) {
+    op->walk([&](tensor::DimOp dimOp) { dimOps.insert(dimOp.getOperation()); });
+  }
+
+  /// Return all tracked tensor::DimOps.
+  SmallVector<tensor::DimOp> getTensorDimOps() {
+    SmallVector<tensor::DimOp> result;
+    for (Operation *op : dimOps) result.push_back(cast<tensor::DimOp>(op));
+    return result;
+  }
+
+ protected:
+  void notifyOperationRemoved(Operation *op) override {
+    IRRewriter::notifyOperationRemoved(op);
+    if (isa<tensor::DimOp>(op)) dimOps.erase(op);
+  }
+
+  void notifyOperationInserted(Operation *op) override {
+    IRRewriter::notifyOperationInserted(op);
+    if (isa<tensor::DimOp>(op)) dimOps.insert(op);
+  }
+
+ private:
+  SmallPtrSet<Operation *, 16> dimOps;
+};
+}  // namespace
+
+/// Simplfy the given tensor::DimOps as much as possible.
+/// * Static dimensions are replaced by constant.
+/// * Dynamic dim ops are pushed as much as possible to the top of the function,
+///   i.e., if the dim of a value is known to be equal to the dim of a value on
+///   the reverse SSA use-def chain, rewrite the value with a dim op of that
+///   value.
+static LogicalResult simplifyDimOps(RewriterBase &rewriter,
```

Comment:
So really this needs to be a fixed point. I am wondering if we really need to do this. IIUC this is just resolving a `tensor.dim` operation in terms of `tensor.dim` of its operands. But that isnt enough in general. You need to do a fixed point. So we could just use `applyPatternAndFoldOpsGreedily`?

Reply:
alright, thanks for the context. Ill look through again. Couple of issues here
1) I think we can resolve the `arith.constant` hoisting issue during conversion of `flow.dispatch.workgroups` from `flow.dispatch.group`. But we can revisit that later.
2) This might be an indication to upstream canonicalization to not hoist the `arith.constant` by default, or at least have a mode where that can be switched off.

Reply:
Btw, neither are blocking. We can go with the local fixed point.